framework/finalizer.go

```diff
@@ -0,0 +1,96 @@
+package framework
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/giantswarm/microerror"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	finalizerName = "operatorkit.giantswarm.io"
+)
+
+type patchSpec struct {
+	Op    string   `json:"op"`
+	Path  string   `json:"path"`
+	Value []string `json:"value"`
+}
+
+func (f *Framework) addFinalizer(obj interface{}) (bool, error) {
+	accessor, err := meta.Accessor(obj)
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+	if accessor.GetDeletionTimestamp() != nil {
+		return true, nil // object has been marked for deletion, we should ignore it.
+	}
+	if containsFinalizer(accessor.GetFinalizers(), finalizerName) {
+		return false, nil // object already has the finalizer.
+	}
+
+	c, err := rest.RESTClientFor(&rest.Config{})
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+	patch := fmt.Sprintf(`{"op":"add","value":%q,"path":"/metadata/finalizers/1"}`, finalizerName)
+	res := c.Patch(types.JSONPatchType).AbsPath(accessor.GetSelfLink()).Body(patch).Do()
+	if res.Error() != nil {
+		return false, microerror.Mask(res.Error())
+	}
+
+	return true, nil
+}
+
+func (f *Framework) removeFinalizer(ctx context.Context, obj interface{}) error {
+	accessor, err := meta.Accessor(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	if !containsFinalizer(accessor.GetFinalizers(), finalizerName) {
+		f.logger.LogCtx(ctx, "function", "removeFinalizer", "level", "warning", "message", "object is missing a finalizer")
+		return nil // object has no finalizer, probably migration.
+	}
+	c, err := rest.RESTClientFor(&rest.Config{})
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	patch := patchSpec{
+		Op:    "replace",
```

Comment:
Why using the `replace` operation here? Shouldn't it be `remove`? Again a test would help.